From ce19921eebcba9024087ea088ffee982000e35a8 Mon Sep 17 00:00:00 2001
From: Michel JAOUEN <michel.jaouen@st.com>
Date: Wed, 15 May 2024 11:26:59 +0200
Subject: [PATCH 0855/1695] drivers: stm32_ipcc: add capabilities ops provided
 by mailbox framework

Since TXF notification cannot be supported when non secure uses ipcc.
A driver configured without TXF interrupt is registered in mailbox
framework with only MBOX_RX_NOTIF_CAP. A panic is done, if a TXF
interrupt is configured for an IPCC version major 2 and minor 0.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 4da53c8f35567f7086ac642af878fe6e1698755c drivers: stm32_ipcc: add capabilities ops provided by mailbox framework]
Change-Id: I32bdeff46faa512e440a79acc6185242d2300caa
Signed-off-by: Michel JAOUEN <michel.jaouen@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/379903
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_ipcc.c | 64 ++++++++++++++++++++++++++++++---------
 1 file changed, 49 insertions(+), 15 deletions(-)

diff --git a/core/drivers/stm32_ipcc.c b/core/drivers/stm32_ipcc.c
index a4023af1d..b49899d2e 100644
--- a/core/drivers/stm32_ipcc.c
+++ b/core/drivers/stm32_ipcc.c
@@ -31,6 +31,7 @@
 #define IPCC_C2PRIVCFGR			U(0x94)
 #define IPCC_C2CIDCFGR			U(0x98)
 #define IPCC_HWCFGR			U(0x3F0)
+#define IPCC_VER			U(0x3F4)
 
 #define IPCC_C1CR			U(0x00)
 #define IPCC_C1MR			U(0x04)
@@ -114,6 +115,7 @@ struct stm32_ipcc_mbx_data {
 	mbox_rx_callback_t rx_cb;
 	mbox_tx_callback_t tx_cb;
 	struct mbox_dev *cb_ctx;
+	uint32_t caps;
 };
 
 struct ipcc_pdata {
@@ -366,7 +368,8 @@ static enum itr_return stm32_ipcc_mbox_txf_isr(struct itr_handler *h)
 			if (data->tx_cb)
 				data->tx_cb(data->cb_ctx, i);
 			/* Disable txf interrupt */
-			ipcc_channel_transmit(pdata->lbase, i, false);
+			if (data->caps & MBOX_TX_NOTIF_CAP)
+				ipcc_channel_transmit(pdata->lbase, i, false);
 		}
 	}
 	return ITRR_HANDLED;
@@ -398,7 +401,8 @@ static TEE_Result stm32_ipcc_mbox_send(const struct mbox_desc *desc,
 	/* Set channel txs Flag */
 	io_write32(pdata->lbase + IPCC_SCR, BIT(id + IPCC_SCR_CH1S_POS));
 	/* Enable channel txf interrupt */
-	ipcc_channel_transmit(pdata->lbase, id, true);
+	if (data->caps & MBOX_TX_NOTIF_CAP)
+		ipcc_channel_transmit(pdata->lbase, id, true);
 
 	return TEE_SUCCESS;
 }
@@ -504,6 +508,14 @@ static TEE_Result stm32_ipcc_mbox_complete(const struct mbox_desc *desc,
 	return TEE_SUCCESS;
 }
 
+static uint32_t stm32_ipcc_mbox_capabilities(const struct mbox_desc *desc)
+{
+	struct ipcc_pdata *pdata = (struct ipcc_pdata *)desc->priv;
+	struct stm32_ipcc_mbx_data *data = &pdata->data;
+
+	return data->caps;
+}
+
 static const struct mbox_ops stm32_ipcc_mbox_ops = {
 	.send = stm32_ipcc_mbox_send,
 	.register_callback = stm32_ipcc_mbox_register_callback,
@@ -511,6 +523,7 @@ static const struct mbox_ops stm32_ipcc_mbox_ops = {
 	.max_channel_count = stm32_ipcc_mbox_channel_count,
 	.channel_interrupt = stm32_ipcc_mbox_enable,
 	.complete = stm32_ipcc_mbox_complete,
+	.capabilities = stm32_ipcc_mbox_capabilities,
 };
 
 static void stm32_ipcc_mbox_init(struct ipcc_pdata *pdata, const void *fdt,
@@ -530,11 +543,32 @@ static void stm32_ipcc_mbox_init(struct ipcc_pdata *pdata, const void *fdt,
 				       &pdata->itr_num[IPCC_ITR_RXO]);
 	if (res)
 		panic();
+	data->caps = MBOX_RX_NOTIF_CAP;
+
 	res = interrupt_dt_get_by_name(fdt, node, "tx",
 				       &pdata->itr_chip[IPCC_ITR_TXF],
 				       &pdata->itr_num[IPCC_ITR_TXF]);
-	if (res)
+	if (!res) {
+		/*
+		 * This support is possible, if non secure is not using ipcc
+		 * channel , this limitation come with channel interrupt mask
+		 * that can not be written in an atomic access , on ip version
+		 * major 2 and minor 0
+		 */
+		uint32_t ver = io_read32(pdata->base + IPCC_VER);
+		uint8_t major = ver >> 4 & 0xf;
+		uint8_t minor = ver & 0xf;
+
+		if (major == 0x2 && minor == 0)
+			panic();
+		data->caps |= MBOX_TX_NOTIF_CAP;
+	} else if (res != TEE_ERROR_GENERIC) {
+		/*
+		 *  TEE_ERROR_GENERIC is return if node is not present any
+		 *  other return value is unexpected
+		 */
 		panic();
+	}
 
 	desc = calloc(1, sizeof(*desc));
 	if (!desc)
@@ -579,20 +613,20 @@ static void stm32_ipcc_mbox_init(struct ipcc_pdata *pdata, const void *fdt,
 					  &pdata->itr[IPCC_ITR_RXO]);
 	if (res)
 		panic("Cannot register ipcc rx handler");
-
-	res = interrupt_alloc_add_handler(pdata->itr_chip[IPCC_ITR_TXF],
-					  pdata->itr_num[IPCC_ITR_TXF],
-					  stm32_ipcc_mbox_txf_isr,
-					  ITRF_TRIGGER_LEVEL, pdata,
-					  &pdata->itr[IPCC_ITR_TXF]);
-	if (res)
-
-		panic("Cannot register ipcc tx handler");
-
+	if (data->caps & MBOX_TX_NOTIF_CAP) {
+		res = interrupt_alloc_add_handler(pdata->itr_chip[IPCC_ITR_TXF],
+						  pdata->itr_num[IPCC_ITR_TXF],
+						  stm32_ipcc_mbox_txf_isr,
+						  ITRF_TRIGGER_LEVEL, pdata,
+						  &pdata->itr[IPCC_ITR_TXF]);
+		if (res)
+			panic("Cannot register ipcc tx handler");
+
+		interrupt_enable(pdata->itr_chip[IPCC_ITR_TXF],
+				 pdata->itr[IPCC_ITR_TXF]->it);
+	}
 	interrupt_enable(pdata->itr_chip[IPCC_ITR_RXO],
 			 pdata->itr[IPCC_ITR_RXO]->it);
-	interrupt_enable(pdata->itr_chip[IPCC_ITR_TXF],
-			 pdata->itr[IPCC_ITR_TXF]->it);
 
 	/* Register device to mailbox framework */
 	if (mbox_dt_register(fdt, node, desc))
-- 
2.47.2

